src/pages/iou/steps/MoneyRequestAmountForm.js

```diff
@@ -0,0 +1,266 @@
+
+    /** Fired when submit button pressed, saves the given amount and navigates to the next page */
+    onSubmitButtonPress: PropTypes.func.isRequired,
+};
+
+const defaultProps = {
+    amount: 0,
+    currency: CONST.CURRENCY.USD,
+    forwardedRef: null,
+    isEditing: false,
+};
+
+/**
+ * Returns the new selection object based on the updated amount's length
+ *
+ * @param {Object} oldSelection
+ * @param {Number} prevLength
+ * @param {Number} newLength
+ * @returns {Object}
+ */
+const getNewSelection = (oldSelection, prevLength, newLength) => {
+    const cursorPosition = oldSelection.end + (newLength - prevLength);
+    return {start: cursorPosition, end: cursorPosition};
+};
+
+const AMOUNT_VIEW_ID = 'amountView';
+const NUM_PAD_CONTAINER_VIEW_ID = 'numPadContainerView';
+const NUM_PAD_VIEW_ID = 'numPadView';
+
+function MoneyRequestAmountForm({amount, currency, isEditing, forwardedRef, onCurrencyButtonPress, onSubmitButtonPress}) {
+    const {translate, toLocaleDigit, fromLocaleDigit, numberFormat} = useLocalize();
+
+    const textInput = useRef(null);
+
+    const selectedAmountAsString = amount ? CurrencyUtils.convertToWholeUnit(currency, amount).toString() : '';
+
+    const [currentAmount, setCurrentAmount] = useState(selectedAmountAsString);
+    const [shouldUpdateSelection, setShouldUpdateSelection] = useState(true);
+
+    const [selection, setSelection] = useState({
+        start: selectedAmountAsString.length,
+        end: selectedAmountAsString.length,
+    });
+
+    /**
+     * Event occurs when a user presses a mouse button over an DOM element.
+     *
+     * @param {Event} event
+     * @param {Array<string>} nativeIds
+     */
+    const onMouseDown = (event, nativeIds) => {
+        const relatedTargetId = lodashGet(event, 'nativeEvent.target.id');
+        if (!_.contains(nativeIds, relatedTargetId)) {
+            return;
+        }
+        event.preventDefault();
+        if (!textInput.current) {
+            return;
+        }
+        if (!textInput.current.isFocused()) {
+            textInput.current.focus();
+        }
+    };
+
+    /**
+     * Convert amount to whole unit and update selection
+     *
+     * @param {String} currencyCode
+     * @param {Number} amountInCurrencyUnits
+     */
+    const saveAmountToState = (currencyCode, amountInCurrencyUnits) => {
+        if (!currencyCode || !amountInCurrencyUnits) {
+            return;
+        }
+        const amountAsStringForState = CurrencyUtils.convertToWholeUnit(currencyCode, amountInCurrencyUnits).toString();
+        setCurrentAmount(amountAsStringForState);
+        setSelection({
+            start: amountAsStringForState.length,
+            end: amountAsStringForState.length,
+        });
+    };
+
+    useEffect(() => {
+        saveAmountToState(currency, amount);
+        // we want to update the state only when the amount is changed
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [amount]);
```

Comment:
**NAB/possible follow up:** 

It seems a bit strange to keep the formatted amount in state. Could the formatted amount also just be computed when the `props.amount` change? That kind of makes more sense to me since:

- `props.amount` will change
- recompute the formatted amount on render (possibly optimized with useMemo if necessary)

Something like that seems like it would avoid an additional render from setting the state immediately each time this prop changes.

Maybe the logic related to the selection stuff makes this more difficult. I would try to compute that stuff as well if possible. I'm not sure there is a clear reason why it needs to be in state.

Reply:
+1 to this comment, we need to keep the amount and currency as separate properties in the transaction object so I think it would make sense to not format the amount til we need to display it

Reply:
@mountiny @marcaaron because of the selection logic I think we should keep it.  we have to keep the value of the text input somehow and it’s easier to keep it that way (as a string) as changing the value will occur more often by changing the input value than by changing the prop. The selection logic is used in both TextInputWithCurrencySymbol and BigNumberPad